src/pykeen/training/lcwa.py

```diff
@@ -237,3 +225,77 @@ def _check_slicing_availability(self, supports_sub_batching: bool):
             report = "This model doesn't support sub-batching and slicing is not" " implemented for this model yet."
         logger.warning(report)
         raise MemoryError("The current model can't be trained on this hardware with these parameters.")
+
+
+# note: we use Tuple[Tensor] here, so we can re-use TensorDataset instead of having to create a custom one
+class SymmetricLCWATrainingLoop(TrainingLoop[Tuple[MappedTriples], Tuple[MappedTriples]]):
+    r"""
+    A "symmetric" LCWA scoring heads *and* tails at once.
+
+    This objective was introduced by [lacroix2018]_ as
+
+    .. math ::
+
+        l_{i,j,k}(X) = - X_{i,j,k} + \log \left(
+            \sum_{k'} \exp(X_{i,j,k′})
+        \right) - X_{k,j+P,i} + \log \left(
+            \sum_{i'} \exp (X_{k, j+P, i'})
+        \right)
+
+
+    which can be seen as a "symmetric LCWA", where for one batch of triples, we score both, heads *and* tails, given
+    the remainder of the triple.
+    """
+
+    # docstr-coverage: inherited
+    def _create_training_data_loader(
+        self, triples_factory: CoreTriplesFactory, sampler: Optional[str], **kwargs
+    ) -> DataLoader[Tuple[MappedTriples]]:  # noqa: D102
+        assert sampler is None
+        return DataLoader(dataset=TensorDataset(triples_factory.mapped_triples), **kwargs)
+
+    # docstr-coverage: inherited
+    def _process_batch(
+        self,
+        batch: Tuple[MappedTriples],
+        start: int,
+        stop: int,
+        label_smoothing: float = 0,
+        slice_size: Optional[int] = None,
+    ) -> torch.FloatTensor:  # noqa: D102
+        # unpack
+        hrt_batch = batch[0]
+        # Send batch to device
+        hrt_batch = hrt_batch[start:stop].to(device=self.model.device)
+        return (
+            # head prediction
+            self.loss.process_lcwa_scores(
+                predictions=self.model.score_h(rt_batch=hrt_batch[:, 1:], slice_size=slice_size, mode=self.mode),
+                # TODO: exploit sparsity
+                # note: this is different to what we do for LCWA, where we collect *all* training entities
+                #   for which the combination is true
```

Comment:
I am not sure whether the name should stay like that, since the targets are different than in the normal LCWA loop.

Reply:
For now, I added a note that the name might change in the future: https://github.com/pykeen/pykeen/pull/1098/commits/91f6bc8f37300d6a96755745836806088805e026